### PR TITLE
Admin username change

### PR DESCRIPTION
**Pull request by jdalsem**
This feature is admin only

username2

### CODE REVIEW

File: views/default/core/settings/account/username.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	'value' => elgg_echo('user:username:change'),
+	'class' => 'elgg-button-action',
+	'data-toggle-selector' => '.account-username-change',
+]);
+
+$body .= elgg_view_field([
+	'#type' => 'text',
+	'#help' => elgg_echo('user:username:help'),
+	'#class' => 'hidden account-username-change',
+	'#data-toggle-slide' => '0',
```
Comment (hypeJunction) [inline views/default/core/settings/account/username.php]: What does this do?

**Review by hypeJunction — COMMENTED** [review]
I don't see a compelling use case to pass more parameters to the field. It opens up doors for plugina to do hacky things and we end up with field mess again. I'd rather we be opinionated.

File: views/default/core/settings/account/username.php (hunk 1 above)
Comment (jdalsem) [inline views/default/core/settings/account/username.php]: it disables the slide animation and just toggles the visibility

**Comment by jdalsem** [comment]
> I don't see a compelling use case to pass more parameters to the field. It opens up doors for plugina to do hacky things and we end up with field mess again. I'd rather we be opinionated.

this way i can not toggle fields visibility for example

File: views/default/core/settings/account/username.php (hunk 1 above)
Comment (hypeJunction) [inline views/default/core/settings/account/username.php]: Ah, the crazy toggler API. Any reason why you wouldn't want to just display the field?

**Comment by hypeJunction** [comment]
Let's rework the toggler to use CSS transitions.

File: views/default/core/settings/account/username.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	'rel' => 'toggle',
+	'value' => elgg_echo('user:username:change'),
+	'class' => 'elgg-button-action',
+	'data-toggle-selector' => '.account-username-change',
+]);
+
+$body .= elgg_view_field([
+	'#type' => 'text',
+	'#help' => elgg_echo('user:username:help'),
+	'#class' => 'hidden account-username-change',
```
Comment (hypeJunction) [inline views/default/core/settings/account/username.php]: We should allow `#id` though, and use that as the toggler selector. Me no like using toggling by class name

Comment (jdalsem) [inline views/default/core/settings/account/username.php]: that is what i am doing here... hiding the button... showing the field

File: views/default/core/settings/account/username.php (hunk 1 above)
Comment (jdalsem) [inline views/default/core/settings/account/username.php]: wanted to make changing the username a more 'special'... like the danger zone (https://github.com/Elgg/Elgg/settings)

don't want an accidental change

Comment (jdalsem) [inline views/default/core/settings/account/username.php]: but if we think admins are sane enough (and they can always change it back)... we could do without the toggler

Comment (hypeJunction) [inline views/default/core/settings/account/username.php]: You could make it `special` by loading a separate form in a lightbox and not mix it with user settings.

**Comment by jdalsem** [comment]
Updated the PR
- no longer does things that toggle

**Comment by jdalsem** [comment]
should i rebase this for 2.x or keep it on master?

**Comment by mrclay** [comment]
I'd prefer a separate action for this as I can imagine sites wanting to limit its use. A lightbox form sounds good.

**Comment by mrclay** [comment]
Store an annotation with JSON: previous username, time changed, user who changed. Firing a hook also good idea. Too bad events don't have params so we could pass this stuff.

**Comment by jdalsem** [comment]
> I'd prefer a separate action for this as I can imagine sites wanting to limit its use.

will be doing this in a separate PR (refactoring more of the usersettings forms)



> Store an annotation with JSON: previous username, time changed, user who changed. Firing a hook also good idea. Too bad events don't have params so we could pass this stuff.

i think all this info is already in system_log

File: engine/lib/user_settings.php [hunk 3]
```diff
@@ -125,10 +125,82 @@ function _elgg_set_user_name() {
...
+	if (!elgg_is_admin_logged_in()) {
+		return;
+	}
+	
+	$user = get_user($user_guid);
+	if (empty($user)) {
+		return;
+	}
+	
+	if (!$user->canEdit()) {
```
Comment (jeabakker) [inline engine/lib/user_settings.php]: what would be the use case that an admin is logged in and he can't edit the user?

File: engine/lib/user_settings.php [hunk 4]
```diff
@@ -125,10 +125,82 @@ function _elgg_set_user_name() {
...
+		return;
+	}
+		
+	if ($user->username == $username) {
```
Comment (jeabakker) [inline engine/lib/user_settings.php]: for PHP reasons ```$user->username === $username```

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/lib/user_settings.php (hunk 4 above)
Comment (jdalsem) [inline engine/lib/user_settings.php]: fixed

File: engine/lib/user_settings.php (hunk 3 above)
Comment (jdalsem) [inline engine/lib/user_settings.php]: fixed

**Review by jeabakker — APPROVED** [review]

**Pull request merged**